File: team_pink/server_f.py
#!/usr/bin/env python3
"""
Process F - Team Pink Worker Server
Worker process that handles queries for Team Pink data subset
"""

import json
import grpc
from concurrent import futures
import sys
import os

# Add proto and common directories to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'proto'))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'common'))

import fire_service_pb2
import fire_service_pb2_grpc
from fire_column_model import FireColumnModel


class FireQueryServiceImpl(fire_service_pb2_grpc.FireQueryServiceServicer):
    """Implementation of FireQueryService for Process F (Worker)"""
    
    def __init__(self, config):
        self.config = config
        self.process_id = config['identity']
        self.role = config['role']
        self.team = config['team']
        print(f"[{self.process_id}] Initialized as {self.role} for Team {self.team}")
        
        # Initialize FireColumnModel with Team Pink data subset
        self.data_model = FireColumnModel()
        data_path = os.path.join(os.path.dirname(__file__), '..', 'data')
        if os.path.exists(data_path):
            # Check if partition is configured
            allowed_dirs = None
            if 'data_partition' in config and config['data_partition'].get('enabled'):
                allowed_dirs = config['data_partition'].get('directories', [])
                print(f"[{self.process_id}] Loading partitioned data from {len(allowed_dirs)} subdirectories...")
            
            self.data_model.read_from_directory(data_path, allowed_dirs)
            print(f"[{self.process_id}] Data model initialized with {self.data_model.measurement_count()} measurements")
        else:
            print(f"[{self.process_id}] Data directory not found: {data_path}")
            print(f"[{self.process_id}] Data model initialized with 0 measurements")
    
    def Query(self, request, context):
        """
        Handle client query request (if called directly)
        Workers typically don't receive direct client queries
        """
        print(f"[{self.process_id}] Received direct query request_id={request.request_id}")
        
        # Workers don't typically receive direct client queries
        chunk = fire_service_pb2.QueryResponseChunk(
            request_id=request.request_id,
            chunk_number=0,
            is_last_chunk=True,
            total_chunks=1,
            total_results=0
        )
        yield chunk
    
    def InternalQuery(self, request, context):
        """
        Handle internal queries from team leader (Process E)
        This is the main method for workers
        """
        print(f"[{self.process_id}] Internal query from {request.requesting_process}")
        print(f"  Request ID: {request.request_id}")
        print(f"  Original request: {request.original_request_id}")
        print(f"  Query type: {request.query_type}")
        
        # Query local FireColumnModel data
        local_measurements = self._query_local_data(request)
        print(f"[{self.process_id}] Found {len(local_measurements)} local measurements")
        
        # Return response with local results (workers don't forward to anyone)
        response = fire_service_pb2.InternalQueryResponse(
            request_id=request.request_id,
            original_request_id=request.original_request_id,
            is_complete=True,
            responding_process=self.process_id
        )
        response.measurements.extend(local_measurements)
        
        print(f"[{self.process_id}] Returning response with {len(response.measurements)} measurements")
        return response
    
    def _query_local_data(self, request):
        """
        Query local FireColumnModel data
        Returns list of FireMeasurement proto messages
        """
        matching_indices = []
        
        if request.HasField('filter'):
            filter_obj = request.filter
            
            # Start with parameter or site filtering (OR logic for multiple parameters)
            if len(filter_obj.parameters) > 0:
                # Handle multiple parameters (OR logic - match any parameter)
                all_param_indices = set()
                for param in filter_obj.parameters:
                    param_indices = self.data_model.get_indices_by_parameter(param)
                    all_param_indices.update(param_indices)
                matching_indices = list(all_param_indices)
            elif len(filter_obj.site_names) > 0:
                # Filter by site name
                site = filter_obj.site_names[0]
                matching_indices = self.data_model.get_indices_by_site(site)
            else:
                # No parameter/site filter - start with all measurements
                matching_indices = list(range(self.data_model.measurement_count()))
            
            # Apply AQI range filter (AND logic - must also match AQI range)
            if filter_obj.min_aqi > 0 or filter_obj.max_aqi > 0:
                filtered_indices = []
                for idx in matching_indices:
                    aqi = self.data_model.aqis[idx]
                    if ((filter_obj.min_aqi == 0 or aqi >= filter_obj.min_aqi) and
                        (filter_obj.max_aqi == 0 or aqi <= filter_obj.max_aqi)):
                        filtered_indices.append(idx)
                matching_indices = filtered_indices
        else:
            # No filter - return all measurements
            matching_indices = list(range(self.data_model.measurement_count()))
        
        # Convert to proto messages
        measurements = []
        for idx in matching_indices:
            measurement = fire_service_pb2.FireMeasurement(
                latitude=self.data_model.latitudes[idx],
                longitude=self.data_model.longitudes[idx],
                datetime=self.data_model.datetimes[idx],
                parameter=self.data_model.parameters[idx],
                concentration=self.data_model.concentrations[idx],
                unit=self.data_model.units[idx],
                raw_concentration=self.data_model.raw_concentrations[idx],
                aqi=self.data_model.aqis[idx],
                category=self.data_model.categories[idx],
                site_name=self.data_model.site_names[idx],
                agency_name=self.data_model.agency_names[idx],
                aqs_code=self.data_model.aqs_codes[idx],
                full_aqs_code=self.data_model.full_aqs_codes[idx]
            )
            measurements.append(measurement)
        
        return measurements
    
    def CancelRequest(self, request, context):
        """Handle request cancellation"""
        print(f"[{self.process_id}] Cancel request_id={request.request_id}")
        
        return fire_service_pb2.StatusResponse(
            request_id=request.request_id,
            status="cancelled",
            chunks_delivered=0,
            total_chunks=0
        )
    
    def GetStatus(self, request, context):
        """Handle status check"""
        print(f"[{self.process_id}] Status request_id={request.request_id}")
        
        return fire_service_pb2.StatusResponse(
            request_id=request.request_id,
            status="pending",
            chunks_delivered=0,
            total_chunks=0
        )
    
    def Notify(self, request, context):
        """Handle notifications from other processes"""
        print(f"[{self.process_id}] Notification from {request.requesting_process}")
        
        return fire_service_pb2.StatusResponse(
            request_id=request.request_id,
            status="acknowledged"
        )


def load_config(config_path):
    """Load configuration from JSON file"""
    with open(config_path, 'r') as f:
        return json.load(f)


def serve(config_path):
    """Start the gRPC server"""
    # Load configuration
    config = load_config(config_path)
    process_id = config['identity']
    hostname = config['hostname']
    port = config['port']
    
    # Create server
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    
    # Add service implementation
    service_impl = FireQueryServiceImpl(config)
    fire_service_pb2_grpc.add_FireQueryServiceServicer_to_server(service_impl, server)
    
    # Bind to address
    server_address = f"{hostname}:{port}"
    server.add_insecure_port(server_address)
    
    # Start server
    server.start()
    print(f"[{process_id}] Server started on {server_address}")
    print(f"[{process_id}] Press Ctrl+C to stop")
    
    # Keep server running
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        print(f"\n[{process_id}] Shutting down...")
        server.stop(0)


if __name__ == '__main__':
    if len(sys.argv) != 2:
        print("Usage: python server_f.py <config_file>")
        print("Example: python server_f.py ../configs/process_f.json")
        sys.exit(1)
    
    config_path = sys.argv[1]
    serve(config_path)

File: team_pink/server_f.cpp
/**
 * @file server_f.cpp
 * @brief Process F - Team Pink Worker Server (C++)
 * 
 * Worker process that handles queries for Team Pink data subset
 */

#include <iostream>
#include <memory>
#include <string>
#include <fstream>
#include <grpcpp/grpcpp.h>
#include <nlohmann/json.hpp>

#include "../proto/fire_service.grpc.pb.h"
#include "FireColumnModel.hpp"

using grpc::Server;
using grpc::ServerBuilder;
using grpc::ServerContext;
using grpc::Status;
using fire_service::FireQueryService;
using fire_service::QueryRequest;
using fire_service::QueryResponseChunk;
using fire_service::InternalQueryRequest;
using fire_service::InternalQueryResponse;
using fire_service::StatusRequest;
using fire_service::StatusResponse;
using fire_service::FireMeasurement;
using json = nlohmann::json;

/**
 * @class FireQueryServiceImpl
 * @brief Implementation of FireQueryService for Process F (Worker)
 */
class FireQueryServiceImpl final : public FireQueryService::Service {
private:
    std::string process_id_;
    std::string role_;
    std::string team_;
    int port_;
    
    // FireColumnModel for local data storage
    FireColumnModel data_model_;

public:
    FireQueryServiceImpl(const json& config) {
        process_id_ = config["identity"];
        role_ = config["role"];
        team_ = config["team"];
        port_ = config["port"];
        
        std::cout << "[" << process_id_ << "] Initialized as " << role_ 
                  << " for Team " << team_ << std::endl;
        
        // Initialize FireColumnModel with Team Pink data subset
        std::cout << "[" << process_id_ << "] Loading data from data/ directory..." << std::endl;
        data_model_.readFromDirectory("data/");
        std::cout << "[" << process_id_ << "] Data model initialized with " 
                  << data_model_.measurementCount() << " measurements" << std::endl;
    }
    
    /**
     * Query RPC - Workers typically don't receive direct client queries
     */
    Status Query(ServerContext* context, 
                 const QueryRequest* request,
                 grpc::ServerWriter<QueryResponseChunk>* writer) override {
        std::cout << "[" << process_id_ << "] Received direct query request_id=" 
                  << request->request_id() << std::endl;
        
        // Return empty response
        QueryResponseChunk chunk;
        chunk.set_request_id(request->request_id());
        chunk.set_chunk_number(0);
        chunk.set_is_last_chunk(true);
        chunk.set_total_chunks(1);
        chunk.set_total_results(0);
        
        writer->Write(chunk);
        return Status::OK;
    }
    
    /**
     * InternalQuery RPC - Main method for workers
     * Receives queries from team leader (Process E)
     */
    Status InternalQuery(ServerContext* context,
                        const InternalQueryRequest* request,
                        InternalQueryResponse* response) override {
        std::cout << "[" << process_id_ << "] Internal query from " 
                  << request->requesting_process() << std::endl;
        std::cout << "  Request ID: " << request->request_id() << std::endl;
        std::cout << "  Original request: " << request->original_request_id() << std::endl;
        std::cout << "  Query type: " << request->query_type() << std::endl;
        
        // Query local FireColumnModel data
        std::vector<std::size_t> matching_indices;
        
        if (request->has_filter()) {
            const auto& filter = request->filter();
            
            // Filter by parameter (PM2.5, PM10, etc.) - use first parameter if specified
            if (filter.parameters_size() > 0) {
                const std::string& param = filter.parameters(0);
                matching_indices = data_model_.getIndicesByParameter(param);
                std::cout << "  Filtered by parameter '" << param 
                         << "': " << matching_indices.size() << " matches" << std::endl;
            }
            // Filter by site name - use first site if specified
            else if (filter.site_names_size() > 0) {
                const std::string& site = filter.site_names(0);
                matching_indices = data_model_.getIndicesBySite(site);
                std::cout << "  Filtered by site '" << site 
                         << "': " << matching_indices.size() << " matches" << std::endl;
            }
            // Filter by AQI range
            else if (filter.min_aqi() > 0 || filter.max_aqi() > 0) {
                const auto& all_aqis = data_model_.aqis();
                for (std::size_t i = 0; i < all_aqis.size(); ++i) {
                    int aqi = all_aqis[i];
                    bool matches = true;
                    
                    if (filter.min_aqi() > 0 && aqi < filter.min_aqi()) {
                        matches = false;
                    }
                    if (filter.max_aqi() > 0 && aqi > filter.max_aqi()) {
                        matches = false;
                    }
                    
                    if (matches) {
                        matching_indices.push_back(i);
                    }
                }
                std::cout << "  Filtered by AQI range: " << matching_indices.size() << " matches" << std::endl;
            }
        } else {
            // No filter - return all measurements
            for (std::size_t i = 0; i < data_model_.measurementCount(); ++i) {
                matching_indices.push_back(i);
            }
        }
        
        // Convert results to FireMeasurement proto messages
        for (std::size_t idx : matching_indices) {
            auto* measurement = response->add_measurements();
            measurement->set_latitude(data_model_.latitudes()[idx]);
            measurement->set_longitude(data_model_.longitudes()[idx]);
            measurement->set_datetime(data_model_.datetimes()[idx]);
            measurement->set_parameter(data_model_.parameters()[idx]);
            measurement->set_concentration(data_model_.concentrations()[idx]);
            measurement->set_unit(data_model_.units()[idx]);
            measurement->set_raw_concentration(data_model_.rawConcentrations()[idx]);
            measurement->set_aqi(data_model_.aqis()[idx]);
            measurement->set_category(data_model_.categories()[idx]);
            measurement->set_site_name(data_model_.siteNames()[idx]);
            measurement->set_agency_name(data_model_.agencyNames()[idx]);
            measurement->set_aqs_code(data_model_.aqsCodes()[idx]);
            measurement->set_full_aqs_code(data_model_.fullAqsCodes()[idx]);
        }
        
        response->set_request_id(request->request_id());
        response->set_original_request_id(request->original_request_id());
        response->set_is_complete(true);
        response->set_responding_process(process_id_);
        
        std::cout << "[" << process_id_ << "] Returning response with " 
                  << response->measurements_size() << " measurements" << std::endl;
        
        return Status::OK;
    }
    
    /**
     * CancelRequest RPC
     */
    Status CancelRequest(ServerContext* context,
                        const StatusRequest* request,
                        StatusResponse* response) override {
        std::cout << "[" << process_id_ << "] Cancel request_id=" 
                  << request->request_id() << std::endl;
        
        response->set_request_id(request->request_id());
        response->set_status("cancelled");
        response->set_chunks_delivered(0);
        response->set_total_chunks(0);
        
        return Status::OK;
    }
    
    /**
     * GetStatus RPC
     */
    Status GetStatus(ServerContext* context,
                    const StatusRequest* request,
                    StatusResponse* response) override {
        std::cout << "[" << process_id_ << "] Status request_id=" 
                  << request->request_id() << std::endl;
        
        response->set_request_id(request->request_id());
        response->set_status("pending");
        response->set_chunks_delivered(0);
        response->set_total_chunks(0);
        
        return Status::OK;
    }
    
    /**
     * Notify RPC
     */
    Status Notify(ServerContext* context,
                 const InternalQueryRequest* request,
                 StatusResponse* response) override {
        std::cout << "[" << process_id_ << "] Notification from " 
                  << request->requesting_process() << std::endl;
        
        response->set_request_id(request->request_id());
        response->set_status("acknowledged");
        
        return Status::OK;
    }
};

/**
 * Load configuration from JSON file
 */
json load_config(const std::string& config_path) {
    std::ifstream file(config_path);
    if (!file.is_open()) {
        throw std::runtime_error("Could not open config file: " + config_path);
    }
    
    json config;
    file >> config;
    return config;
}

/**
 * Start the gRPC server
 */
void RunServer(const std::string& config_path) {
    // Load configuration
    json config = load_config(config_path);
    std::string process_id = config["identity"];
    std::string hostname = config["hostname"];
    int port = config["port"];
    
    // Build server address
    std::string server_address = hostname + ":" + std::to_string(port);
    
    // Create service implementation
    FireQueryServiceImpl service(config);
    
    // Build and start server
    ServerBuilder builder;
    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
    builder.RegisterService(&service);
    
    std::unique_ptr<Server> server(builder.BuildAndStart());
    std::cout << "[" << process_id << "] Server started on " << server_address << std::endl;
    std::cout << "[" << process_id << "] Press Ctrl+C to stop" << std::endl;
    
    // Wait for server shutdown
    server->Wait();
}

int main(int argc, char** argv) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <config_file>" << std::endl;
        std::cerr << "Example: " << argv[0] << " ../configs/process_f.json" << std::endl;
        return 1;
    }
    
    try {
        RunServer(argv[1]);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

